File: delta/src/directory.js
// PII-довідник `handle → {name, email, lang}` — M4 (docs/specs/
// 260809-delta-app.md, «Обсяг M4», п.1; конституція п.8 «Мультиюзер на
// ~20»: «Ідентичність = handle, PII поза git (`.mt/directory.json`,
// адмін-екран у додатку)», і docs/specs/260714-cognitive-delegation.md,
// п.1 «Ідентичність = handle, PII — поза git»). У файлах вузлів (mandates.yaml,
// decision-request) НІКОЛИ немає email/повного імені — лише handle; це
// відображення живе окремо, ПОЗА git (той самий рівень приватності, що
// `device_key.json`/`knowledge.json`), і ЩЕ й у самому `mandatesDir`
// (на відміну від них — файл-сусідів `config.json` одного пристрою), бо
// довідник спільний для ВСІХ пристроїв воркспейсу (адмінка редагує один
// файл, усі учасники бачать ті самі display-імена).
//
// **PII не комітиться:** `<mandatesDir>/.mt/directory.json` — той самий
// шлях, що згадує конституція; корінь репо ігнорує `**/.mt/directory.json`
// (.gitignore) саме тому, що тестові/демо-воркспейси цього застосунку
// живуть під `delta/` — фікстура з РЕАЛЬНИМИ іменами/email НЕ повинна
// потрапити в git навіть випадково. Приклад формату — `.example`-файл
// (`src/tests/fixtures/directory.example.json`), не сам `directory.json`.
//
// Схема: плаский обʼєкт `{ [handle]: { name: string|null, email: string|null,
// lang: string|null } }` — О(1) lookup за handle (на відміну від масиву
// `device-registry.json`, де записи звіряються за ДВОМА полями одночасно
// — тут ключ один, handle, і саме він первинний).

/**
 * @returns {object} порожній довідник
 */
export function emptyDirectory() {
  return {}
}

/**
 * @param {unknown} raw сирий запис одного handle з розпарсеного JSON
 * @returns {{name: string|null, email: string|null, lang: string|null}} нормалізований запис
 */
function normalizeEntry(raw) {
  const entry = raw && typeof raw === 'object' ? raw : {}
  return {
    name: typeof entry.name === 'string' && entry.name.trim() ? entry.name.trim() : null,
    email: typeof entry.email === 'string' && entry.email.trim() ? entry.email.trim() : null,
    lang: typeof entry.lang === 'string' && entry.lang.trim() ? entry.lang.trim() : null
  }
}

/**
 * Розбирає сирий текст `directory.json` — відсутній/битий файл повертає
 * порожній обʼєкт (не кидає), той самий інваріант, що `device-registry.js`/
 * `knowledge.js`: відсутність довідника — стан «ще ніхто не заповнив
 * display-імена», не помилка.
 * @param {string|null|undefined} text сирий вміст файлу
 * @returns {object} нормалізований довідник
 */
export function parseDirectory(text) {
  if (!text) return {}
  try {
    const parsed = JSON.parse(text)
    if (!parsed || typeof parsed !== 'object' || Array.isArray(parsed)) return {}
    const result = {}
    for (const [handle, raw] of Object.entries(parsed)) result[handle] = normalizeEntry(raw)
    return result
  } catch {
    return {}
  }
}

/**
 * Серіалізує довідник у pretty-print JSON з кінцевим переносом рядка,
 * ключі відсортовані (стабільний diff у git — довідник комітиться).
 * @param {object} entries довідник
 * @returns {string} текст файлу
 */
export function formatDirectory(entries) {
  const sorted = {}
  for (const handle of Object.keys(entries).toSorted()) sorted[handle] = entries[handle]
  return `${JSON.stringify(sorted, null, 2)}\n`
}

/**
 * Записує (чи оновлює) запис одного handle — pure-функція (повертає новий
 * обʼєкт, не мутує вхідний), part update (лише передані поля змінюються,
 * решта лишається — read-merge-write на рівні одного запису).
 * @param {object} entries поточний довідник
 * @param {string} handle handle запису
 * @param {{name?: string, email?: string, lang?: string}} patch поля для оновлення
 * @returns {object} новий довідник із оновленим записом
 */
export function setDirectoryEntry(entries, handle, patch) {
  const current = entries[handle] ?? { name: null, email: null, lang: null }
  const merged = normalizeEntry({ ...current, ...patch })
  return { ...entries, [handle]: merged }
}

/**
 * Display-імʼя для handle — «підставляється в усі екрани (черга, карта,
 * Довіряю) з фолбеком на handle» (M4 п.1). Немає запису/імені — сам handle
 * (ніколи не порожній рядок для валідного handle), `null`/`undefined`
 * handle — `null` (немає кого показувати).
 * @param {object} entries довідник
 * @param {string|null|undefined} handle handle для пошуку
 * @returns {string|null} display-імʼя, або handle як фолбек
 */
export function displayName(entries, handle) {
  if (!handle) return null
  return entries[handle]?.name ?? handle
}

File: delta/crates/delta-core/src/directory.rs
//! PII-довідник `handle → {name, email, lang}` — порт `delta/src/directory.js`
//! (M4). Живе В `mandatesDir`, ПОЗА git (`.gitignore`).

use std::collections::BTreeMap;

#[derive(Debug, Clone, PartialEq, Default, serde::Serialize, serde::Deserialize)]
pub struct DirectoryEntry {
    pub name: Option<String>,
    pub email: Option<String>,
    pub lang: Option<String>,
}

fn normalize_str(v: Option<&str>) -> Option<String> {
    v.map(str::trim)
        .filter(|s| !s.is_empty())
        .map(str::to_string)
}

fn normalize_entry(raw: &serde_json::Value) -> DirectoryEntry {
    DirectoryEntry {
        name: normalize_str(raw.get("name").and_then(|v| v.as_str())),
        email: normalize_str(raw.get("email").and_then(|v| v.as_str())),
        lang: normalize_str(raw.get("lang").and_then(|v| v.as_str())),
    }
}

pub type Directory = BTreeMap<String, DirectoryEntry>;

pub fn empty_directory() -> Directory {
    BTreeMap::new()
}

/// Розбирає сирий текст `directory.json` — відсутній/битий файл повертає
/// порожній довідник (не кидає) — `directory.js: parseDirectory`.
pub fn parse_directory(text: Option<&str>) -> Directory {
    let Some(text) = text else {
        return empty_directory();
    };
    if text.trim().is_empty() {
        return empty_directory();
    }
    let Ok(parsed) = serde_json::from_str::<serde_json::Value>(text) else {
        return empty_directory();
    };
    let Some(obj) = parsed.as_object() else {
        return empty_directory();
    };
    obj.iter()
        .map(|(handle, raw)| (handle.clone(), normalize_entry(raw)))
        .collect()
}

/// Серіалізує довідник у pretty-print JSON, ключі відсортовані (`directory.js:
/// formatDirectory`) — `BTreeMap` уже дає стабільний ітераційний порядок.
pub fn format_directory(entries: &Directory) -> String {
    format!(
        "{}\n",
        serde_json::to_string_pretty(entries).expect("Directory серіалізується без помилок")
    )
}

pub struct DirectoryPatch {
    pub name: Option<String>,
    pub email: Option<String>,
    pub lang: Option<String>,
}

/// Записує (чи оновлює) запис одного handle — part update (лише передані
/// поля змінюються) — `directory.js: setDirectoryEntry`.
pub fn set_directory_entry(entries: &Directory, handle: &str, patch: DirectoryPatch) -> Directory {
    let current = entries.get(handle).cloned().unwrap_or_default();
    let merged = DirectoryEntry {
        name: patch.name.or(current.name),
        email: patch.email.or(current.email),
        lang: patch.lang.or(current.lang),
    };
    let mut next = entries.clone();
    next.insert(handle.to_string(), merged);
    next
}

/// Display-імʼя для handle — фолбек на сам handle (`directory.js:
/// displayName`).
pub fn display_name<'a>(entries: &'a Directory, handle: Option<&'a str>) -> Option<&'a str> {
    let handle = handle?;
    Some(
        entries
            .get(handle)
            .and_then(|e| e.name.as_deref())
            .unwrap_or(handle),
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_missing_or_corrupt_is_empty() {
        assert!(parse_directory(None).is_empty());
        assert!(parse_directory(Some("not json")).is_empty());
        assert!(parse_directory(Some("[]")).is_empty());
    }

    #[test]
    fn set_and_display_name() {
        let entries = empty_directory();
        let updated = set_directory_entry(
            &entries,
            "olena",
            DirectoryPatch {
                name: Some("Olena K.".into()),
                email: None,
                lang: None,
            },
        );
        assert_eq!(display_name(&updated, Some("olena")), Some("Olena K."));
    }

    #[test]
    fn display_name_falls_back_to_handle_when_no_entry() {
        let entries = empty_directory();
        assert_eq!(display_name(&entries, Some("ghost")), Some("ghost"));
        assert_eq!(display_name(&entries, None), None);
    }

    #[test]
    fn partial_update_preserves_other_fields() {
        let entries = empty_directory();
        let step1 = set_directory_entry(
            &entries,
            "olena",
            DirectoryPatch {
                name: Some("Olena".into()),
                email: Some("o@x.com".into()),
                lang: None,
            },
        );
        let step2 = set_directory_entry(
            &step1,
            "olena",
            DirectoryPatch {
                name: None,
                email: None,
                lang: Some("uk".into()),
            },
        );
        let entry = &step2["olena"];
        assert_eq!(entry.name.as_deref(), Some("Olena"));
        assert_eq!(entry.email.as_deref(), Some("o@x.com"));
        assert_eq!(entry.lang.as_deref(), Some("uk"));
    }

    #[test]
    fn format_round_trips() {
        let entries = empty_directory();
        let updated = set_directory_entry(
            &entries,
            "olena",
            DirectoryPatch {
                name: Some("Olena".into()),
                email: None,
                lang: None,
            },
        );
        let text = format_directory(&updated);
        assert_eq!(parse_directory(Some(&text)), updated);
    }
}
